### DISCUSSION TITLE
Support Apache Tika for RAG text extraction

### DISCUSSION THREAD

**Discussion by nickovs** [open-webui/open-webui#3475] [General]
Currently, document attachments for RAG are parsed using by [selecting](https://github.com/open-webui/open-webui/blob/9e4dd4b86f77653b59dc1fa97fe3f72e8252b359/backend/apps/rag/main.py#L970-L1062) from a [grab-bag](https://github.com/open-webui/open-webui/blob/9e4dd4b86f77653b59dc1fa97fe3f72e8252b359/backend/apps/rag/main.py#L21-L37) of document loaders from the [LangChain community set](https://python.langchain.com/v0.1/docs/modules/data_connection/document_loaders/). While this avoids the need for external services, the supported file type set is small, the results are not always high quality in terms of output order and spacing, and it doesn't support valuable features such as OCR.

It would be great if Open WebUI optionally allowed use of [Apache Tika](https://tika.apache.org) as an alternative way of parsing attachments.

Tika has mature support for parsing [hundreds of different document formats](https://tika.apache.org/3.0.0-BETA/formats.html), which would greatly expand the set of documents that could be passed in to Open WebUI. It also has integrated support for applying OCR to embedded images, so for instance text extraction from a PDF that is made up of scans of pages "just works".

Importantly, useful installations of Tika are available as completely [self-contained Docker images](https://hub.docker.com/r/apache/tika) with a REST interface, including versions with bundled Tesseract OCR, making deployment as part of a `docker-compose.yml` very easy.

Supporting Tika would involve providing a configuration option to let the admin set a Tika service URL and then fixing up `backend/apps/rag/main.py` to offload most of the text extraction work to Tika if this variable is set.

**Comment by tjbck** [comment]
I believe the new Filter function should enable this use case. I'd love to collaborate on this if you're interested!

**Reply by nickovs** [reply]
Right now the current filter implementation doesn't allow this for several reasons:

1. Firstly, `pipelines` filters don't have access to the raw attachment file since it's not accessible through the `/documents/doc` endpoint and your `pipelines` instance might be running in a different container which won't have direct access to the uploads directory.
1. Even if the previous issue was fixed, _right now_ calls out to filters are performed using `requests.post()`, which blocks, and in order to even get to the existing data from the `/documents/doc` endpoint, you have to be able to start another request to the (blocked) Open WebUI backend, so you end up in a deadlock.
1. It's not the **right place** to put it, unless the existing RAG file extraction is going to be moved to a whole new type of pipeline filter, since (a) this should really be a replacement to the existing text extraction function and (b) by the time even a priority 0 filter is called the existing text extraction code has already been run and the text already indexed.

There is also a separate issue in that the current chat body that is passed to the filter has separate `message` and `docs` lists, and you can't straightforwardly work out the order in which these should be interleaved in a multi-message chat.

My feeling is that:
1. should probably be addresses as a new feature request for a new `/documents/...` endpoint, 
2. is a bug, and calls out to pipelines should be switched to `aiohttp` because right now one stalled pipeline filter can hang the whole Open WebUI system
3. is the real reason why this should be implemented in the core backend, rather than as a filter.

That said, I'm happy to hear thoughts from others.

**Reply by tjbck** [reply]
I meant the filter “function” not the filter “pipeline”, we just introduced the new functions feature with 0.3.6 alongside Files API, hope that clarifies!

**Reply by nickovs** [reply]
@tjbck Thanks for the clarification. Can you point me to some documentation (or at least specific source code) for this new functionality?

**Reply by nickovs** [reply]
Having run some test, for the general task of just getting text and metadata out of any file you care to through at it, here is all the code that you need to make an asynchronous call to Tika and return the contents:
```python
async def tika_text(data, mine_type=None):
    headers = {"Content-Type": mime_type} if mime_type else {}
    async with aiohttp.ClientSession() as session:
        async with session.put(TIKA_BASE_URL + "tika/text", data=data, headers=headers) as response:
            metadata = await response.json()
            text = metadata['X-TIKA:content']
    return text, metadata
```

**Comment by nickovs** [comment]
I have a working implementation of Tika integration, although right now its server address is hard-wired because I have no skills in the JavaScript/Svelte space. Basically I created a simple Tika loader class in `backend/apps/rag/main.py` that looks like this:

```python
class TikaLoader:
    def __init__(self, file_path, mime_type=None):
        self.file_path = file_path
        self.mime_type = mime_type

    def load(self) -> List[Document]:
        with open(self.file_path, "rb") as f:
            data = f.read()

            if self.mime_type is not None:
                headers = {"Content-Type": self.mime_type}
            else:
                headers = {}

            endpoint = TIKA_SERVER_URL + ("" if TIKA_SERVER_URL[-1] == "/" else "/") + "tika/text"

            r = requests.put(endpoint, data=data, headers=headers)

            if r.ok:
                raw_metadata = r.json()
                text = raw_metadata.get("X-TIKA:content", "<No text content found>")

                return [Document(page_content=text, metadata=headers)]
            else:
                raise Exception("Error calling Tika")
```

I then updated [`get_loader(...)`](https://github.com/open-webui/open-webui/blob/24b638fcdea17fb8621e0391150bee852cf74a38/backend/apps/rag/main.py#L988) to preface the large `if...elif...` ladder with:
```python
    if USE_TIKA:
        if file_ext in known_source_ext or (
                file_content_type and file_content_type.find("text/") >= 0
        ):
            loader = TextLoader(file_path, autodetect_encoding=True)
        else:
            loader = TikaLoader(file_path, file_content_type)
    else:
        ...
```

This is currently enabled with:
```python
USE_TIKA = True
TIKA_SERVER_URL = "http://tika:9998"
```
but of course this needs be replaced with appropriate user configuration.

I'm currently running this with Tika running in an adjacent Docker container by sticking these two lines into my `docker-compose.yaml`:
```yaml
  tika:
    image: apache/tika:latest-full
```
The document parsing works perfectly on pretty much anything you can throw at it, from scans of pages to calendar `.ics` files to old format PowerPoint.

**Reply by medrpa** [reply]
Thank for this information. Can you please make a tutorial on that?
I am not understanding properly. I am not using docker so how the code will be modify.

**Comment by nickovs** [comment]
@medrpa Support for Tika was added to the main line of OpenWebUI in PR #3582, which was merged back in July. You should just be able to set this up in the RAG config panel now.
